src/main/java/build/buf/protovalidate/internal/evaluator/EvaluatorBuilder.java

```diff
@@ -78,358 +78,404 @@ public EvaluatorBuilder(Env env, boolean disableLazy) {
    * unknown, returns an evaluator that always throws a {@link CompilationException}.
    */
   public Evaluator load(Descriptor desc) throws CompilationException {
-    if (disableLazy) {
-      return loadDescriptor(desc);
-    } else {
-      return loadOrBuildDescriptor(desc);
+    Evaluator evaluator = evaluatorCache.get(desc);
+    if (evaluator == null && disableLazy) {
+      return new UnknownDescriptorEvaluator(desc);
     }
+    return build(desc);
   }
 
   /**
    * Either returns a memoized {@link Evaluator} for the given descriptor, or lazily constructs a
    * new one.
    */
   private Evaluator build(Descriptor desc) throws CompilationException {
-    Evaluator eval = evaluatorMap.get(desc);
+    Evaluator eval = evaluatorCache.get(desc);
     if (eval != null) {
       return eval;
     }
-    MessageEvaluator msgEval = new MessageEvaluator();
-    evaluatorMap.put(desc, msgEval);
-    buildMessage(desc, msgEval);
-    return msgEval;
-  }
-
-  private void buildMessage(Descriptor desc, MessageEvaluator msgEval) throws CompilationException {
-    try {
-      DynamicMessage defaultInstance =
-          DynamicMessage.newBuilder(desc).mergeFrom(new byte[0], EXTENSION_REGISTRY).buildPartial();
-      Descriptor descriptor = defaultInstance.getDescriptorForType();
-      MessageConstraints msgConstraints =
-          resolver.resolveMessageConstraints(descriptor, EXTENSION_REGISTRY);
-      if (msgConstraints.getDisabled()) {
-        return;
+    synchronized (this) {
```

Comment:
This is the key change - if we have a cache miss we'll check again (in case another thread beat us in the meantime) and if that fails, we'll rebuild a brand new cache map from the current one and the descriptor.